### PR TITLE
fix: Unable to build the image of schematic-api

### LINKED ISSUE #1573: [Bug] Fix the task `build-image` of the project `schematic-api`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What projects are you seeing the problem on?

Schematic

### Current behavior

Building the Docker image of Schematic REST API fails with the following error:

```console
$ nx build-image schematic-api
...

#10 26.86 
#10 ERROR: executor failed running [/bin/bash -euxo pipefail -c pip install poetry   && poetry config --local virtualenvs.create false   && poetry install --with prod --no-root --no-interaction --no-ansi   && pip cache purge]: exit code: 1
------
 > [6/9] RUN pip install poetry   && poetry config --local virtualenvs.create false   && poetry install --with prod --no-root --no-interaction --no-ansi   && pip cache purge:
#10 26.86         2│ import json
#10 26.86         3│ import os
#10 26.86         4│ from pathlib import Path
#10 26.86         5│ 
#10 26.86     →   6│ from virtualenv.create.describe import Python2Supports
#10 26.86         7│ from virtualenv.create.via_global_ref.builtin.ref import PathRefToDest
#10 26.86         8│ from virtualenv.info import IS_ZIPAPP
#10 26.86         9│ from virtualenv.util.zipapp import read as read_from_zipapp
#10 26.86        10│ 
#10 26.86 
------
ERROR: failed to solve: executor failed running [/bin/bash -euxo pipefail -c pip install poetry   && poetry config --local virtualenvs.create false   && poetry install --with prod --no-root --no-interaction --no-ansi   && pip cache purge]: exit code: 1
>  Nx Container  Removing temp folder /tmp/docker-build-push-xpjXi3
 >  NX   buildx failed with: ERROR: failed to solve: executor failed running [/bin/bash -euxo pipefail -c pip install poetry   && poetry config --local virtualenvs.create false   && poetry install --with prod --no-root --no-interaction --no-ansi   && pip cache purge]: exit code: 1
   Pass --verbose to see the stacktrace.

 —————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Ran target build-image for project schematic-api (32s)
```

### Expected behavior

The command should create the image successfully.

### Anything else?

_No response_

### Commit ID

4edc585

### Are you developing inside the dev container?

- [X] I am using the dev container

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by tschaffter**
Fixes #1573

## Description

The project `schematic-api` was no longer working. One solution was to update from Python 3.9.2 to Python 3.11.0. I decided to implement this solution as this gave me the opportunity to explore how to enable projects in this monorepo to use different versions of Python if needed.

## Changelog

- Update the project `schematic-api` to use Python 3.11.0
- Fix the task `build-image` of the project `schematic-api` by upgrading to Python 3.11.0
- Add the workspace command `schematic-build-images` to build all the images of the Schematic projects.

## Related PRs

- #1576
- #1582

## Notes

- Installing a Python environment may take more than 1 min (depends on compute resources and internet bandwidth). The time required to perform this action is larger than the time required to download a slightly larger devcontainer image that is preloaded with the environment. For this reason, the Sage devcontainer now accepts multiple Python environments to be included in its image (up to a limit). 

    ```console
    $ time pyenv install 3.11.0
    ...
    real    1m31.234s
    ```

## How to test this PR

- Checkout the branch of this PR
- This branch uses a recently created devcontainer. After checking out the branch, rebuild and reopen the devcontainer.
- Run the commands listed in the section Preview (see below)

### Prepare the python environment

```console
nx prepare-python schematic-api
```

### Start the Schematic REST API in development mode

```console
nx serve schematic-api
```

The Swagger UI page is now available on `localhost:7080`

image

### Build the Docker image

```console
nx build-image schematic-api
```

### Start the Schematic REST API with Docker

```console
nx serve-detach schematic-api
```

The Swagger UI page is now available on `localhost:7080`

### Run the unit tests

```console
nx test schematic-api
```

### Build all the Schematic images

```console
$ schematic-build-images 

    ✔  nx run schematic-api-description:build  [local cache]
    ✔  nx run schematic-api-docs:build  [local cache]
    ✔  nx run schematic-api-docs:build-image (3s)
    ✔  nx run schematic-api:build-image (53s)
```

### CODE REVIEW

File: apps/schematic/api/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9.2-slim-buster
+FROM python:3.11.0-slim-buster
```
Comment (linglp) [inline apps/schematic/api/Dockerfile:1]: hmm I am not sure if matters here. But schematic supports python 3.9 and python3.10 for now.

File: apps/schematic/api/project.json [hunk 2]
```diff
@@ -15,10 +15,13 @@
       "executor": "nx:run-commands",
```
Comment (linglp) [inline apps/schematic/api/project.json:15]: hmm what is project.json used for?

File: apps/schematic/api/Dockerfile (hunk 1 above)
Comment (tschaffter) [reply apps/schematic/api/Dockerfile:1]: See the Description of this PR that I just added above. The version of Python can be changed in the future if needed.

File: apps/schematic/api/project.json (hunk 2 above)
Comment (tschaffter) [reply apps/schematic/api/project.json:15]: The projects in this monorepo are managed with a tool called `nx`. If a folder in this monorepo has a file `project.json`, then Nx will detect this folder as a "project". The file `project.json` defines the tasks that are implemented for a given project. These tasks are then executed with Nx with the command `nx <task> <project>` or `nx run <project>:<task>`. E.g., `nx prepare-python schematic-api` (or `nx run schematic-api:prepare-python`), `nx build schematic-api-description` (or `nx run schematic-api-description:build`), etc.

**Comment by tschaffter** [comment]
Hey @linglp , I'm sorry if you received notification about this PR over the weekend. I wanted to make sure that the issue I found was fixed before you started the week on Monday.

The PR should now work fine if you try to check it out and run the command that are listed in the section above named "Preview". Keep an eye open for a message in the terminal that tells you to rebuild your devcontainer (see below).

image

VS Code may also detect that the branch of this PR needs a different devcontainer, in which case it may show a dialog to invite you to rebuild the container.

**Comment by linglp** [comment]
@tschaffter No worries Thomas. I figured out the rebuild part on my own. I could run the schematic APIs based on the instructions.  :) Thanks for fixing the bug.

**Review by linglp — APPROVED** [review]

File: apps/schematic/api/project.json (hunk 2 above)
Comment (linglp) [reply apps/schematic/api/project.json:15]: Yes. This makes sense. Thanks Thomas. I was able to run schematic-api-description too

**Pull request merged**